### DISCUSSION TITLE
hibernate-validator can't find built-in constraint

### DISCUSSION THREAD

**Discussion by ghost** [quarkusio/quarkus#22776] [Q&A] [answered]
I have a Quarkus app using `resteasy-reactive` and I'm trying to introduce bean validation to validate the request body. I've decorated the mapped POJOs with constraint annotations and added the `@Valid` annotation to the corresponding argument on the request method. I'm getting exceptions like this:

```Request failed: javax.validation.UnexpectedTypeException: HV000030: No validator could be found for constraint 'javax.validation.constraints.NotNull' validating type 'java.lang.String'. Check configuration for 'method.argument.field'```

Just wanted to check if this is a known issue before I create a reproducer...

[a comment by an automated account is not shown]

**Comment by gsmet** [comment]
Hmmm, no, it should work out of the box. Are your mapped POJOs in the Jandex index (i.e. are they part of your main app or if not do you add them to the index)?

**Reply by ghost** [reply]
The POJOs are in a dependency. I'll check to make sure they are properly indexed, but I didn't think of that as a problem since it is picking up the constraint - it just can't find a suitable validator. Does Quarkus prune out validators based on the known constrain annotations?

**Reply by gsmet** [reply]
> Does Quarkus prune out validators based on the known constrain annotations?

More precisely, it only registers the validators for the constraints it finds, built-in validators included. But what's bugging me a bit is that if it didn't find anything, it shouldn't apply the constraints if the ValidatorFactory created by Quarkus is used.

Do you build your own ValidatorFactory?

**Reply by ghost** [reply]
> Do you build your own ValidatorFactory?

No, and Interestingly, if I inject `Validator` and validate manually in the body of the request method, it passes with no violations.

**Reply by ghost** [reply]
Anyway, the issue was indeed caused by the lack of the Jandex index. Running the packaged version worked as expected, but I was initially using `clean compile quarkus:dev` in a multi-module project, so it wasn't building the indexes. Using `clean compile process-classes quarkus:dev` did the trick. 👍

**Reply by gsmet** [reply]
@evc-erik yeah so that's what I suspected.

Your problem with the index is something I noticed the other day. I think we should probably adjust `quarkus:dev` so that it also builds the index (we also made it do the `compile`). /cc @aloubyansky @glefloch WDYT?

Anyway, it's far better/optimized to use the VF provided by Quarkus (and it will work in native), you can either inject a VF or a Validator and do things from there.

**Reply by glefloch** [reply]
Does adding a empty `beans.xml` file in the resource directory of the module containing the pogo fix the issue?

**Reply by gsmet** [reply]
Yeah, my question was more about the fact that if you do `mvn clean quarkus:dev` or the Gradle equivalent probably in a multi-module project, it does a `compile` but it does not build the Jandex indexes of the submodules if they are built with the Maven plugin. So you end up with no indexes when doing that.

**Reply by aloubyansky** [reply]
That would make sense.

**Reply by aloubyansky** [reply]
Actually @evc-erik how does your jandex plugin look like?

**Reply by glefloch** [reply]
Right, does it mean we want to index all sub-modules of the application ?

**Reply by gsmet** [reply]
> https://github.com/quarkusio/quarkus/pull/15560/files

No but if the Jandex plugin is declared is a submodule, it should be executed when doing `mvn clean quarkus:dev` (or the Gradle equivalent). Last time I checked what Maven is doing when we do `mvn clean quarkus:dev`, it didn't.

**Reply by aloubyansky** [reply]
That makes sense. Although if you specify `process-classes`, you don't have to specify `compile`, since `compile` is a dependency of `process-classes`, so `clean process-classes quarkus:dev` should be enough.

**Reply by ghost** [reply]
> Actually @evc-erik how does your jandex plugin look like?

The sub-module just has this:
```xml
  <build>
    <plugins>
      <plugin>
        <groupId>org.jboss.jandex</groupId>
        <artifactId>jandex-maven-plugin</artifactId>
        <version>1.2.1</version>
        <executions>
          <execution>
            <goals>
              <goal>jandex</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
```

**Reply by ghost** [reply]
> `clean process-classes quarkus:dev` should be enough.

Sure, but remembering `process-classes` is the tricky part 😛

Also, I'm wondering what happens during development reloads. Will new indexes get created if classes are changed/added?

**Reply by ghost** [reply]
> Does adding a empty `beans.xml` file in the resource directory of the module containing the pogo fix the issue?

No, it does not.

**Reply by aloubyansky** [reply]
> > Does adding a empty `beans.xml` file in the resource directory of the module containing the pogo fix the issue?
> 
> No, it does not.

Are you sure about that? That would be a bug.

**Reply by aloubyansky** [reply]
Just to clarify, it should be `META-INF/beans.xml`

**Reply by ghost** [reply]
> Just to clarify, it should be `META-INF/beans.xml`

Ah, right. I've been out of practice! It does work if I add `META-INF/beans.xml`.

**Comment by aloubyansky** [comment]
> Also, I'm wondering what happens during development reloads. Will new indexes get created if classes are changed/added?

Yes, in fact, in dev mode and tests we are re-generating the Jandex indexes for local projects anyway to make sure they are up-to-date.

**Comment by ghost** [comment, marked as answer]
Okay, so just to wrap this up, the solution to my problem is to ensure that the module containing the validated POJOs is properly available to the [bean discovery](https://quarkus.io/guides/cdi-reference#bean_discovery) mechanism.

A related question about whether `quarkuss:dev` should imply `process-classes` in addition to `compile` remains open. (FWIW, I think it should!)

Thanks for the help everyone!

**Reply by aloubyansky** [reply]
@evc-erik would you mind creating an issue to cover `process-classes` in `quarkus:dev`? Thanks.

**Reply by ghost** [reply]
@aloubyansky I created #22807
